docs/src/formula.md

```diff
@@ -14,167 +14,184 @@ fields with possibly heterogeneous types.  One of the primary goals of
-```
+julia> data[:y] = X*β_true .+ ϵ
 
-The `@formula` macro re-writes the formula expression `y ~ 1 + (a+b)*c` as a
-call to the `Formula` constructor.  The arguments of the constructor correspond
-to the fields of the `Formula` struct, which are, in order:
+julia> fit(LinearModel, @formula(y ~ 1 + a*b), data)
+StatsModels.TableRegressionModel{LinearModel{LmResp{Array{Float64,1}},DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}},Array{Float64,2}}
 
-* `ex_orig`: the original expression `:(y ~ 1+(a+b)*c)`
-* `ex`: the parsed expression `:(y ~ 1+a+b+a&c+b&c)`
-* `lhs`: the left-hand side `:y`
-* `rhs`: the right-hand side `:(1+a+b+a&c+b&c)`
+y (continuous) ~ 1 + a (continuous) + b (4 levels): DummyCoding(3) + a (continuous) & b (4 levels): DummyCoding(3)
 
-```@docs
-Formula
-dropterm
+Coefficients:
+             Estimate Std.Error t value Pr(>|t|)
+(Intercept)  0.945011 0.0443292  21.318   <1e-36
+a             2.04038 0.0781928 26.0942   <1e-43
+b: e          3.06689 0.0585607 52.3711   <1e-69
+b: f          4.03094 0.0620725 64.9393   <1e-77
+b: g          5.06553 0.0656504 77.1592   <1e-84
+a & b: e      5.90041  0.101969 57.8648   <1e-73
+a & b: f       6.9842  0.105653 66.1048   <1e-78
+a & b: g      7.93213  0.108797 72.9077   <1e-82
 ```
 
-## The `ModelFrame` and `ModelMatrix` types
-
-The main use of `Formula`s is for fitting statistical models based on tabular
-data.  From the user's perspective, this is done by `fit` methods that take a
-`Formula` and a `DataFrame` instead of numeric matrices.
-
-Internally, this is accomplished in three stages:
-
-1. The `Formula` is parsed into [`Terms`](@ref).
-2. The `Terms` and the data source are wrapped in a [`ModelFrame`](@ref).
-3. A numeric [`ModelMatrix`](@ref) is generated from the `ModelFrame` and passed to the
-   model's `fit` method.
-
-```@docs
-ModelFrame
-ModelMatrix
-Terms
-```
+Internally, this is accomplished in three steps:
+
+1. The expression passed to `@formula` is lowered to term constructors combined
+   by `~`, `+`, and `&`, which evaluate to create terms for the whole formula
+   and any interaction terms.
+2. A schema is extracted from the data, which determines whether each variable
+   is continuous or categorical and extracts the summary statistics of each
+   variable (mean/variance/min/max or unique levels respectively).  This schema
+   is then _applied_ to the formula with `apply_schema(term, schema,
+   ::Type{Model})`, which returns a new formula with each placeholder `Term`
+   replaced with a concrete `ContinuousTerm` or `CategoricalTerm` as
+   appropriate.  This is also the stage where any custom syntax is applied (see
+   [the section on extending the `@formula`
+   DSL](@ref Internals-and-extending-the-formula-DSL) for more details)
+3. Numeric arrays are generated for the response and predictors from the full
+   table using `model_cols(term, data)`.
```

Comment:
Question about the terminology: isn't it slightly inconsistent to use an imperative verb for `apply_schema`, but only nouns for `model_cols`?

Reply:
perhaps but I couldn't think of something imperative and concise.  also there's teh convention that a noun named function means "make one of these from your arguments" so that's kind of what I was going for.  same with `schema`.  Do you have another suggestion?

Reply:
Naming is hard. After thinking more about it, I think one of the reasons I find the names inconsistent is that `apply_schema` describes the input (schema), but not the output (contrary to `model_cols`, as you note). Can we find a name describing the result of `apply_schema`? `model_terms`?

Reply:
my original idea was `instantiate` but I thought that was maybe too vague